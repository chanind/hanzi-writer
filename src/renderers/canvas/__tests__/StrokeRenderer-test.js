const yi = require('hanzi-writer-data/一.json');
const StrokeRenderer = require('../StrokeRenderer');
const parseCharData = require('../../../parseCharData');


const char = parseCharData('一', yi);

describe('StrokeRenderer', () => {

  let ctx;

  beforeEach(() => {
    ctx = document.createElement('canvas').getContext('2d');
  });

  it('renders a path and clipPath', () => {
    const props = {
      strokeColor: {r: 12, g: 101, b: 20, a: 0.3},
      radicalColor: null,
      strokeWidth: 2,
      opacity: 0.7,
      displayPortion: 0.4,
    };
    const renderer = new StrokeRenderer(char.strokes[0]);
    renderer.render(ctx, props);

    expect(ctx.__getEvents()).toMatchSnapshot();
  });

  it('works without using Path2D if needed', () => {
    const props = {
      strokeColor: {r: 12, g: 101, b: 20, a: 1},
      radicalColor: null,
      strokeWidth: 2,
      opacity: 0.7,
      displayPortion: 0.4,
    };
    const renderer = new StrokeRenderer(char.strokes[0], false);
    renderer.render(ctx, props);

    expect(ctx.__getEvents()).toMatchSnapshot();
  });

  it('skips rendering if opacity is close to 0', () => {
    const props = {
      strokeColor: {r: 12, g: 101, b: 20, a: 0.3},
      radicalColor: null,
      strokeWidth: 2,
      opacity: 0.01,
      displayPortion: 0.4,
    };
    const renderer = new StrokeRenderer(char.strokes[0]);
    renderer.render(ctx, props);
    expect(ctx.__getEvents()).toEqual([]);
  });

});
