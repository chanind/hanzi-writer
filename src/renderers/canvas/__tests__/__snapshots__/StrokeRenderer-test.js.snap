// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StrokeRenderer renders a path and clipPath 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": 0.7,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  Object {
    "props": Object {
      "value": "rgba(12, 101, 20, 0.3)",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": "rgba(12, 101, 20, 0.3)",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": 200,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineJoin",
  },
  Object {
    "props": Object {
      "value": Array [
        910.9281897070704,
        910.9281897070704,
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setLineDash",
  },
  Object {
    "props": Object {
      "value": 546.010356910418,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineDashOffset",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 25,
      "y": 421,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 193,
      "y": 372,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 417,
      "y": 402,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 827,
      "y": 434,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 920,
      "y": 401,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 25,
            "y": 421,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 193,
            "y": 372,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 417,
            "y": 402,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 827,
            "y": 434,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 920,
            "y": 401,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;

exports[`StrokeRenderer works without using Path2D if needed 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 518,
      "y": 382,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "cpx": 572,
      "cpy": 385,
      "x": 623,
      "y": 389,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 758,
      "cpy": 399,
      "x": 900,
      "y": 383,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 928,
      "cpy": 379,
      "x": 935,
      "y": 390,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 944,
      "cpy": 405,
      "x": 930,
      "y": 419,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 896,
      "cpy": 452,
      "x": 845,
      "y": 475,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 829,
      "cpy": 482,
      "x": 798,
      "y": 473,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 723,
      "cpy": 460,
      "x": 480,
      "y": 434,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 180,
      "cpy": 409,
      "x": 137,
      "y": 408,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 130,
      "cpy": 408,
      "x": 124,
      "y": 408,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 108,
      "cpy": 408,
      "x": 106,
      "y": 395,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 105,
      "cpy": 380,
      "x": 127,
      "y": 363,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 146,
      "cpy": 348,
      "x": 183,
      "y": 334,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 195,
      "cpy": 330,
      "x": 216,
      "y": 338,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 232,
      "cpy": 344,
      "x": 306,
      "y": 354,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "cpx": 400,
      "cpy": 373,
      "x": 518,
      "y": 382,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "quadraticCurveTo",
  },
  Object {
    "props": Object {
      "value": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 518,
            "y": 382,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 518,
            "y": 382,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": 0.7,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  Object {
    "props": Object {
      "value": "#0c6514",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": "#0c6514",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": 200,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineJoin",
  },
  Object {
    "props": Object {
      "value": Array [
        910.9281897070704,
        910.9281897070704,
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setLineDash",
  },
  Object {
    "props": Object {
      "value": 546.010356910418,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineDashOffset",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 25,
      "y": 421,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 193,
      "y": 372,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 417,
      "y": 402,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 827,
      "y": 434,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 920,
      "y": 401,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 25,
            "y": 421,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 193,
            "y": 372,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 417,
            "y": 402,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 827,
            "y": 434,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 920,
            "y": 401,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;
